File: dotnet/src/PuzzleSolver.BreadthFirstSearch/BreadthFirstSearch.cs
using PuzzleSolver.Search;

namespace PuzzleSolver.BreadthFirstSearch;

public class BreadthFirstSearch<TState, TStateTraversalDescription> : ISearch<TState, TStateTraversalDescription>
    where TStateTraversalDescription : class
{
    private readonly Predicate<TState> _successIndicator;
    private readonly Func<StateTraversal<TState, TStateTraversalDescription>, IEnumerable<StateTraversal<TState, TStateTraversalDescription>>> _nextStateGenerator;
    private readonly Queue<StateTraversal<TState, TStateTraversalDescription>> _frontier;
    private readonly HashSet<TState> _explored;
    private readonly IEqualityComparer<TState> _stateComparer;

    public BreadthFirstSearch(
        Predicate<TState> successIndicator,
        Func<StateTraversal<TState, TStateTraversalDescription>, IEnumerable<StateTraversal<TState, TStateTraversalDescription>>> nextStateGenerator,
        IEqualityComparer<TState> stateComparer)
    {
        _successIndicator = successIndicator;
        _nextStateGenerator = nextStateGenerator;
        _frontier = new Queue<StateTraversal<TState, TStateTraversalDescription>>();
        _explored = new HashSet<TState>(stateComparer);
        _stateComparer = stateComparer;
    }

    public IEnumerable<StateTraversal<TState, TStateTraversalDescription>> Execute(TState initialState)
    {
        _frontier.Enqueue(new StateTraversal<TState, TStateTraversalDescription>(
            null, 
            null, 
            initialState));
        var maxLength = 0;

        while (_frontier.TryDequeue(out var candidate))
        {
            _explored.Add(candidate.Child);

            if (_successIndicator(candidate.Child))
            {
                yield return candidate;
            }

            foreach (var additionalCandidate in _nextStateGenerator(candidate))
            {
                if (!_frontier.Select(f => f.Child).Any(s => _stateComparer.Equals(s, additionalCandidate.Child)) 
                    && !_explored.Contains(additionalCandidate.Child))
                {
                    _frontier.Enqueue(additionalCandidate);
                    if (_frontier.Count > maxLength)
                    {
                        maxLength = _frontier.Count;
                    }
                }
            }
        }
    }
}

File: python/breadth_first_search.py
from collections import deque
from collections.abc import Callable, Iterable
from typing import TypeVar

from predicate import Predicate
from state_traversal import StateTraversal

TState = TypeVar("TState")
TStateTraversalDescription = TypeVar("TStateTraversalDescription")


class BreadthFirstSearch[TState, TStateTraversalDescription]:
    def __init__(
            self,
            success_indicator: Predicate[TState],
            next_state_generator: Callable[
                [StateTraversal[TState, TStateTraversalDescription]],
                Iterable[StateTraversal[TState, TStateTraversalDescription]]]) -> None:
        self._success_indicator = success_indicator
        self._next_state_generator = next_state_generator
        self._candidates: deque[StateTraversal[TState, TStateTraversalDescription]] \
            = deque()
        self._visited: set[TState] = set()

    def execute(
            self,
            initial_state: TState
    ) -> Iterable[StateTraversal[TState, TStateTraversalDescription]]:
        """Performs the search starting from the initial state."""

        self._candidates.append(
            StateTraversal(
                parent=None,
                description=None,
                child=initial_state))

        while self._candidates:
            candidate = self._candidates.popleft()
            self._visited.add(candidate.child)

            if self._success_indicator(candidate.child):
                yield candidate

            for additional_candidate in self._next_state_generator(candidate):
                if additional_candidate.child not in self._visited:
                    self._candidates.append(additional_candidate)
